qemu/tests/qemu_guest_agent_update.py

```diff
@@ -0,0 +1,157 @@
+
+                # get the previous qemu-ga pkg
+                qga_html = params["qga_html"]
+                qga_url = params["qga_url"]
+                gagent_html_download_cmd = params["gagent_html_download_cmd"]
+                process.system(gagent_html_download_cmd, login_timeout)
+
+                with open(qga_html, "r") as f:
+                    lines = f.readlines()
+                list_qga = []
+                for line in lines:
+                    if "qemu-ga-win" in line:
+                        list_qga.append(line)
+                tgt_line = list_qga[-2]
+                # qemu-ga-win-7.4.4-1
+                pattern = re.compile(r"qemu-ga-win-\d\.\d\.\d-\d")
+                ret = pattern.findall(tgt_line)[0]
+
+                # https://fedorapeople.org/groups/virt/virtio-win/direct-downloads/
+                #   archive-qemu-ga/qemu-ga-win-7.4.4-1/qemu-ga-x64.msi
+                qga_url_pre = r"%s/%s/%s" % (qga_url, ret, qemu_ga_pkg_pre)
+                gagent_pre_download_cmd = "wget %s -O /tmp/qemu-ga-pre.msi" % qga_url_pre
+                process.system(gagent_pre_download_cmd, login_timeout)
+                self.vm.copy_files_to("/tmp/qemu-ga-pre.msi", self.gagent_guest_dir)
+            else:
+                test.error("Only support 'url' and 'virtio-win' method to "
+                           "download qga installer now.")
+
+        params["start_vm"] = "yes"
+        login_timeout = float(params.get("login_timeout", 360))
+        dst_path_lat = r"%s\%s" % (self.gagent_guest_dir, "qemu-ga-lat.msi")
+        dst_path_pre = r"%s\%s" % (self.gagent_guest_dir, "qemu-ga-pre.msi")
+        qemu_ga_pkg_pre = params["qemu_ga_pkg_pre"]
+        install_cmd = params["gagent_install_cmd"]
+        uninstall_cmd = params["gagent_uninstall_cmd"]
+
+        error_context.context("Boot up vm.", logging.info)
+        env_process.preprocess_vm(test, params, env, params["main_vm"])
+        vm = self.env.get_vm(params["main_vm"])
+        vm.verify_alive()
+        self.vm = vm
+        session = self._get_session(params, self.vm)
+
+        error_context.context("Create the previous and latest qemu-ga pkg "
+                              "in C: volume.", logging.info)
+        create_dst_path()
+
+        error_context.context("Install the previous qemu-ga in guest.",
+                              logging.info)
+        if self._check_ga_pkg(session, params["gagent_pkg_check_cmd"]):
+            error_context.context("Uninstall the latest one.", logging.info)
+            self.gagent_uninstall_cmd = uninstall_cmd % dst_path_lat
+            self.gagent_uninstall(session, self.vm)
+        self.gagent_install_cmd = install_cmd % dst_path_pre
+        self.gagent_install(session, self.vm)
+
+        error_context.context("Update qemu-ga to the current one.", logging.info)
+        self.gagent_install_cmd = install_cmd % dst_path_lat
+        try:
+            self.gagent_install(session, self.vm)
```

Comment:
If  gagent_install() failed, it will run test.fail in gagent_install. Then the following except will not be executed.